### ISSUE TITLE
af::select() and af::replace() inconsistent and incorrect behavior across backends

### ISSUE THREAD

**Issue by yshady** [arrayfire/arrayfire#1249]
I have been struggling with those functions lately, and have managed to summarize my problems in the following snippet:

```
dim4 size = dim4(4,4,4,1);
array SMask = constant(0, size,b8);
SMask(seq(size[0]/2,end),span,span) = 1;
array O = af::complex(constant(20,size,f32), constant(10,size,f32));
array oldO = O.copy();

O = select(SMask, oldO-O*0.9, O);		//Option 1
//replace(O, !SMask, oldO-O*0.9);		//Option 2
//O(SMask) = oldO(SMask)-O(SMask)*0.9;	//Option 3
	
af_print(O);
```

## CPU Backend
Works as expected for all options (1,2,3), and gives this (correct) output.
```
O
[4 4 4 1]
(20.0000,10.0000) (20.0000,10.0000) (20.0000,10.0000) (20.0000,10.0000) 
(20.0000,10.0000) (20.0000,10.0000) (20.0000,10.0000) (20.0000,10.0000) 
(2.0000,1.0000) (2.0000,1.0000) (2.0000,1.0000) (2.0000,1.0000) 
(2.0000,1.0000) (2.0000,1.0000) (2.0000,1.0000) (2.0000,1.0000) 

(20.0000,10.0000) (20.0000,10.0000) (20.0000,10.0000) (20.0000,10.0000) 
(20.0000,10.0000) (20.0000,10.0000) (20.0000,10.0000) (20.0000,10.0000) 
(2.0000,1.0000) (2.0000,1.0000) (2.0000,1.0000) (2.0000,1.0000) 
(2.0000,1.0000) (2.0000,1.0000) (2.0000,1.0000) (2.0000,1.0000) 

(20.0000,10.0000) (20.0000,10.0000) (20.0000,10.0000) (20.0000,10.0000) 
(20.0000,10.0000) (20.0000,10.0000) (20.0000,10.0000) (20.0000,10.0000) 
(2.0000,1.0000) (2.0000,1.0000) (2.0000,1.0000) (2.0000,1.0000) 
(2.0000,1.0000) (2.0000,1.0000) (2.0000,1.0000) (2.0000,1.0000) 

(20.0000,10.0000) (20.0000,10.0000) (20.0000,10.0000) (20.0000,10.0000) 
(20.0000,10.0000) (20.0000,10.0000) (20.0000,10.0000) (20.0000,10.0000) 
(2.0000,1.0000) (2.0000,1.0000) (2.0000,1.0000) (2.0000,1.0000) 
(2.0000,1.0000) (2.0000,1.0000) (2.0000,1.0000) (2.0000,1.0000)
```

## CUDA Backend
Both Option 1 and Option 2 give incorrect results (in different ways). Option 3 works (same output as the CPU backend) but is 2x slower than Option 1 and 2.
### Option 1
```
O
[4 4 4 1]
         (20.0000,10.0000)          (20.0000,10.0000)          (20.0000,10.0000)          (20.0000,10.0000) 
         (20.0000,10.0000)          (20.0000,10.0000)          (20.0000,10.0000)          (20.0000,10.0000) 
         (2.0000,1.0000)          (2.0000,1.0000)          (2.0000,1.0000)          (2.0000,1.0000) 
         (2.0000,1.0000)          (2.0000,1.0000)          (2.0000,1.0000)          (2.0000,1.0000) 

         (0.0000,0.0000)          (0.0000,0.0000)          (0.0000,0.0000)          (0.0000,0.0000) 
         (0.0000,0.0000)          (0.0000,0.0000)          (0.0000,0.0000)          (0.0000,0.0000) 
         (0.0000,0.0000)          (0.0000,0.0000)          (0.0000,0.0000)          (0.0000,0.0000) 
         (0.0000,0.0000)          (0.0000,0.0000)          (0.0000,0.0000)          (0.0000,0.0000) 

         (0.0000,0.0000)          (0.0000,0.0000)          (0.0000,0.0000)          (0.0000,0.0000) 
         (0.0000,0.0000)          (0.0000,0.0000)          (0.0000,0.0000)          (0.0000,0.0000) 
         (0.0000,0.0000)          (0.0000,0.0000)          (0.0000,0.0000)          (0.0000,0.0000) 
         (0.0000,0.0000)          (0.0000,0.0000)          (0.0000,0.0000)          (0.0000,0.0000) 

         (0.0000,0.0000)          (0.0000,0.0000)          (0.0000,0.0000)          (0.0000,0.0000) 
         (0.0000,0.0000)          (0.0000,0.0000)          (0.0000,0.0000)          (0.0000,0.0000) 
         (0.0000,0.0000)          (0.0000,0.0000)          (0.0000,0.0000)          (0.0000,0.0000) 
         (0.0000,0.0000)          (0.0000,0.0000)          (0.0000,0.0000)          (0.0000,0.0000)
```
### Option 2
```
(20.0000,10.0000)          (20.0000,10.0000)          (20.0000,10.0000)          (20.0000,10.0000) 
         (20.0000,10.0000)          (20.0000,10.0000)          (20.0000,10.0000)          (20.0000,10.0000) 
         (2.0000,1.0000)          (2.0000,1.0000)          (2.0000,1.0000)          (2.0000,1.0000) 
         (2.0000,1.0000)          (2.0000,1.0000)          (2.0000,1.0000)          (2.0000,1.0000) 

         (20.0000,10.0000)          (20.0000,10.0000)          (20.0000,10.0000)          (20.0000,10.0000) 
         (20.0000,10.0000)          (20.0000,10.0000)          (20.0000,10.0000)          (20.0000,10.0000) 
         (20.0000,10.0000)          (20.0000,10.0000)          (20.0000,10.0000)          (20.0000,10.0000) 
         (20.0000,10.0000)          (20.0000,10.0000)          (20.0000,10.0000)          (20.0000,10.0000) 

         (20.0000,10.0000)          (20.0000,10.0000)          (20.0000,10.0000)          (20.0000,10.0000) 
         (20.0000,10.0000)          (20.0000,10.0000)          (20.0000,10.0000)          (20.0000,10.0000) 
         (20.0000,10.0000)          (20.0000,10.0000)          (20.0000,10.0000)          (20.0000,10.0000) 
         (20.0000,10.0000)          (20.0000,10.0000)          (20.0000,10.0000)          (20.0000,10.0000) 

         (20.0000,10.0000)          (20.0000,10.0000)          (20.0000,10.0000)          (20.0000,10.0000) 
         (20.0000,10.0000)          (20.0000,10.0000)          (20.0000,10.0000)          (20.0000,10.0000) 
         (20.0000,10.0000)          (20.0000,10.0000)          (20.0000,10.0000)          (20.0000,10.0000) 
         (20.0000,10.0000)          (20.0000,10.0000)          (20.0000,10.0000)          (20.0000,10.0000)
```

## OpenCL Backend
Option 1 crashes in print(). Option 2 and 3 behave the same way as the CUDA backend.
### Option 1
```
O
[4 4 4 1]
terminate called after throwing an instance of 'af::exception'
  what():  ArrayFire Exception (Internal error:998):
In function print
In file src/api/c/print.cpp:87



In function void af::print(const char*, const af::array&)
In file src/api/cpp/util.cpp:21
Aborted (core dumped)
```

**Comment by pavanky** [comment]
yikes that is bad. Thanks for bringing this to our attention. I will try to fix it ASAP.

**Comment by pavanky** [comment]
@yshady This is fixed in PR #1251. It should be in the devel branch by tomorrow.

Having said that, I think the most efficient way to implement what you are doing is to do the following:

```
O -= SMask * 0.9
```

`select` and `replace` are a little bit heavy and should ideally only be used when you expect the data to have `NaNs`.
